Applications/SpreadSheet/SpreadSheet.h

```diff
@@ -0,0 +1,209 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/HashMap.h>
+#include <AK/HashTable.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <AK/Traits.h>
+#include <AK/Types.h>
+#include <AK/WeakPtr.h>
+#include <AK/Weakable.h>
+#include <LibCore/Object.h>
+#include <LibJS/Interpreter.h>
+
+namespace SpreadSheet {
+
+struct Position {
+    String column;
+    size_t row { 0 };
+
+    bool operator==(const Position& other) const
+    {
+        return row == other.row && column == other.column;
+    }
+
+    bool operator!=(const Position& other) const
+    {
+        return !(other == *this);
+    }
+};
+
+class Sheet;
+
+struct Cell : public Weakable<Cell> {
+    Cell(String data, WeakPtr<Sheet> sheet)
+        : dirty(false)
+        , data(move(data))
+        , kind(LiteralString)
+        , sheet(sheet)
+    {
+    }
+
+    bool dirty { false };
+    bool evaluated_externally { false };
+    String data;
+    JS::Value evaluated_data;
+
+    enum Kind {
+        LiteralString,
+        Formula,
+    } kind { LiteralString };
+
+    WeakPtr<Sheet> sheet;
+    Vector<WeakPtr<Cell>> referencees;
+
+    void reference_from(Cell*);
+
+    void set_data(String new_data)
+    {
+        if (data == new_data)
+            return;
+
+        if (new_data.starts_with("=")) {
+            new_data = new_data.substring(1, new_data.length() - 1);
+            kind = Formula;
+        } else {
+            kind = LiteralString;
+        }
+
+        data = move(new_data);
+        dirty = true;
+        evaluated_externally = false;
+    }
+
+    void set_data(JS::Value new_data)
+    {
+        dirty = true;
+        evaluated_externally = true;
+
+        StringBuilder builder;
+
+        builder.append("=");
+        builder.append(new_data.to_string_without_side_effects());
+        data = builder.build();
+
+        evaluated_data = move(new_data);
+    }
+
+    String source() const;
+
+    JS::Value js_data();
+
+    void update(Badge<Sheet>) { update_data(); }
+    void update();
```

Comment:
What is the difference between these two? From just looking at the API I would think "oh, well if I'm not in `Sheet`, I can still update, I just don't pass a badge"

Reply:
The difference is that `update()` goes through `Sheet` as well.
This ensures that there is no equirecursion.